Removing pipeline register type and consolidating to array of struct …

reecewayt opened the pull request:
@nkanderson  Here are the changes we discussed during our white board session. I decided to go with smart pointers in the pipeline array to, hopefully, make our lives easier and prevent memory issues. Let me know if you have questions.

nkanderson commented on src/functional_simulator.cpp:
```diff
@@ -12,59 +15,136 @@ FunctionalSimulator::FunctionalSimulator(RegisterFile* rf, Stats* st, IMemoryPar
+    advancePipeline();
+}
+
+bool FunctionalSimulator::detectHazards() {
+    // If forwarding is enabled, fewer stalls are needed
+    if (forward || isStageEmpty(PipelineStage::DECODE)) {
+        // TODO: Implement hazard detection with forwarding
+        return false;
+    } else {
+        // Simple hazard detection for no forwarding:
+        // Check if instruction in decode stage depends on result of instruction in execute or
+        // memory stage
+        return false;
+    }
+    return false;
+}
```
I was thinking this would detect hazards, but not deal with forwarding logic. It could just indicate whether there is a data dependency, and elsewhere we could have the logic to decide how to deal with that dependency, i.e. whether to stall or forward.

But if you could maybe lay its usage out in pseudo-code or a basic sequence of method calls, I might have a better understanding of how this method is used alongside the others.

reecewayt replied:
This won't do anything with forwarding. It only detects hazards but hazards need to be evaluated differently depending on if forwarding is enabled or not. This is just a place holder for now.

nkanderson replied:
A placeholder sounds fine, I just wanted to clarify what this returns. I think if this returns `false` when there is a data hazard but forwarding is enabled, how will we know that we need to forward a value? I might not be understanding the intention here though, which is why I was wondering about its intended usage in context.

nkanderson commented on tests/functional_simulator/functional_simulator_tests.cpp:
```diff
@@ -54,67 +54,62 @@ class FunctionalSimulatorTest : public ::testing::Test {
+    EXPECT_EQ(sim->getStall(), 0);
 }
 
-/*
-// The following or something similar may be used to test individual methods
-// with mocked memory parser methods like readInstruction
-// Requires addition of `using ::testing::Return;` above
-TEST_F(FunctionalSimulatorTest, FetchReadsInstructionFromMemory) {
-    EXPECT_CALL(mem, readInstruction(0)).Times(1).WillOnce(Return(0x040103E8));
-    sim->instructionFetch();  // Assuming it internally calls readInstruction
```
We may still want this mocking example.

nkanderson reviewed (commented):
Hey, I think a lot of this looks good. One option would be to split off the struct into a separate PR that could be merged once we settle on whether or not to include the `pc` and `branch_taken`, and to have the other methods and logic in a separate PR. I'm not saying all the other methods need to be finalized before we merge, but I do think there's some discussion and possibly modifications. So we can keep it all in this PR, but we might take a little longer to get it merged.

reecewayt commented on src/functional_simulator.cpp:
```diff
@@ -12,59 +15,136 @@ FunctionalSimulator::FunctionalSimulator(RegisterFile* rf, Stats* st, IMemoryPar
+}
+
+void FunctionalSimulator::cycle() {
+    // Execute stages in reverse order since each stage depends on the previous one
+    // TODO: Might need to add other control logic here
+    writeBack();
+    memory();
+    execute();
+    instructionDecode();
+    instructionFetch();
```
Again just a place holder, execution order will depend on how we implement the pipeline. Do we execute stages then shift, or shift then execute stages. I think it should work either way but we have to consider order of operations nonetheless.

reecewayt commented on include/functional_simulator.h:
```diff
@@ -99,45 +132,70 @@ class FunctionalSimulator {
+        EXECUTE = 2,
+        MEMORY = 3,
+        WRITEBACK = 4,
+        NUM_STAGES = 5
```
Sounds good i'll move it.

reecewayt commented on include/functional_simulator.h:
```diff
@@ -2,28 +2,60 @@
+    uint32_t alu_result;   ///< Result of ALU operation or effective address
+    uint32_t memory_data;  ///< Data read from memory (if applicable)
+    std::optional<uint8_t> dest_reg;  ///< Destination register (if any)
+    bool branch_taken;                ///< Whether a branch is taken
```
I think it makes more sense to pair with the instruction that causes the branch.

reecewayt commented on include/functional_simulator.h:
```diff
@@ -2,28 +2,60 @@
-    std::optional<uint8_t> wb_reg;  ///< Destination register number (std::nullopt if N/A)
+struct PipelineStageData {
+    std::unique_ptr<Instruction> instruction;  ///< Instruction object
+    uint32_t pc;           ///< Program counter value when instruction was fetched
+                           ///< TODO: I think we need to store PC at the time of the instruction
+                           ///< to accurately calculate PC relative addressing
```
Here's how I thought about it. Lets say a branch instruction gets fetch at PC 0x1000 then the pipeline fetches two more instructions before branch taken is evaluated. The PC is now 0x1008. So do we calculate the branch address as 0x1000 + offset, or 0x1008 + offset. The photo you linked is from a multi-cycle CPU which might be causing confusion. A pipelined CPU is shown below where you can see the effective address is the PC Branch calculated from the ALU. Which actually shows PC + 4 + Immediate as the effective address during a branch taken. So the correct calculation is 0x1000 + 4 + Immediate. 
![image](https://github.com/user-attachments/assets/91ae19d9-cb15-494c-a7b3-8afa3870776a)

nkanderson replied:
Hmm, gotcha, since my screenshot was from the 5d pipelining lecture, I thought it was just a simplified diagram without the pipeline registers. I see a `PCPlus4E` as well as `PCPlus4D` and `PCPlus4F` - I'm not sure I have the context to understand all the notation in use there. But I see your point, that the PC is being passed from register to register. I think given the project spec wording, it would probably be best to confirm.

reecewayt commented:
Alright I think I addressed your concerns and questions. I say we just get it merged asap and start working on the pipeline stages. We should honestly be trying to finish this up so we can move on to testing the pipeline with programs from Truong's compiler. Even just running those simulations will give us ideas on areas where we messed up, probably in a more holistic sense then trying to run unit tests and figure out the perfect simulator class.
